### PR TITLE
Gobierto Visualizations / Vue Table Component

### LINKED ISSUE #3596: Vue table component
- Create/improve a Vue table component from what we have in Plans: 
  imagen
  - Orderable columns, ASC DESC
  - Can select which columns to show with the right top link
  - Can specify the order of columns (is the same for the table and for the selector). This may be implied by the order of the columns provided to the component
  - Can specify defaults: visible columns, column which is ordered
  - The value of a column might be a link
  - Numeric fields should be right-aligned
- Implement component in Contracts (Home, Contracts tab)
- Implement component in Subvenciones

### PR DESCRIPTION
**Pull request by jorgeatgu**
Closes #3596 

## :v: What does this PR do?

- Create a table as a new `Vue` component
- Implement the new table in Contracts and Subsidies.
- Remove old components, Table and TableRow.

~~⚠️ Warning this PR introduces `jest`, we need research how to` import Vue-directives/filters` to avoid errors. For example, we need to [import i18n again](https://github.com/PopulateTools/gobierto/compare/3596_vue_table_component?expand=1#diff-6e44e909b696e1f297971ccc74e7cfdbb1ea5c91597a2b0eb33b2370dae866f7R34-R39) to pass the tests without errors.~~

## :mag: How should this be manually tested?

[Contracts Summary](https://getafe.gobify.net/visualizaciones/contratos/)
[Contracts Show](https://getafe.gobify.net/visualizaciones/contratos/adjudicaciones)
[Contracts Assignee](https://getafe.gobify.net/visualizaciones/contratos/adjudicaciones/1057964)
[Subsidies Summary](https://getafe.gobify.net/visualizaciones/subvenciones/)
[Subsidies Show](https://getafe.gobify.net/visualizaciones/subvenciones/subvenciones)

### After this PR

Screen Recording 2021-01-22 at 17 26 35 2021-01-22 17_32_47

### CODE REVIEW

File: app/assets/stylesheets/comp-vue-table.scss [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  width: 100%;
+  margin: 0;
+  padding: 0;
+
+  &.gobierto-table-scroll {
+    table {
+      display: block;
+      overflow-x: scroll;
+    }
+  }
```
Comment (Crashillo) [inline app/assets/stylesheets/comp-vue-table.scss]: Avoid unnecessary nesting

File: app/assets/stylesheets/comp-vue-table.scss [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    font-size: $f8;
+    cursor: pointer;
+    text-align: left;
+    border: 0;
+
+    &:last-child {
+      .sort-icon {
+        margin-right: .5rem;
+      }
+    }
```
Comment (Crashillo) [inline app/assets/stylesheets/comp-vue-table.scss]: Same

File: app/assets/stylesheets/comp-vue-table.scss [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    display: flex;
+    align-items: center;
+    padding-right: .5rem;
+
+    > :not(:last-child) {
+      margin-right: 1rem;
+      white-space: nowrap;
+      overflow: hidden;
+      text-overflow: ellipsis;
+      max-width: 35vw;
```
Comment (Crashillo) [inline app/assets/stylesheets/comp-vue-table.scss]: Too narrow for mobile, isn't it?

File: app/assets/stylesheets/comp-vue-table.scss [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    &-term {
+      text-transform: uppercase;
+      font-size: $f5;
+      text-decoration: none;
+      color: var(--color-base);
+      white-space: nowrap;
+      overflow: hidden;
+      text-overflow: ellipsis;
+      max-width: 40vw;
```
Comment (Crashillo) [inline app/assets/stylesheets/comp-vue-table.scss]: Same

File: app/assets/stylesheets/comp-vue-table.scss [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        white-space: nowrap;
+        margin-right: .5rem;
+      }
+    }
+
+    &__icon {
+      &.fas {
+        margin-right: .5rem;
+      }
+    }
```
Comment (Crashillo) [inline app/assets/stylesheets/comp-vue-table.scss]: No nesting, and don't use font-awesome classes, here you may use `i`

File: app/assets/stylesheets/comp-vue-table.scss [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+    &__content {
+      position: absolute;
+      right: 0;
+      top: 100%;
+      background-color: $color_main_negative;
+      padding: 2em;
+      box-sizing: border-box;
+      width: 25vw;
```
Comment (Crashillo) [inline app/assets/stylesheets/comp-vue-table.scss]: Mobile?

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 7]
```diff
@@ -0,0 +1,240 @@
+<template>
+  <div
+    data-testid="table-component"
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: What for?

File: app/javascript/lib/commons/setup-test.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import $ from "jquery";
```
Comment (Crashillo) [inline app/javascript/lib/commons/setup-test.js]: Remove this file

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+                >
+                  <span>{{ item[field] }}</span>
+                </td>
+              </template>
+            </template>
+          </router-link>
+        </template>
+      </tbody>
+    </table>
+    <Pagination
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: Pagination should be an `slot`, you're enforcing always to have a pagination component

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                </td>
+              </template>
+              <template v-else-if="type === 'link'">
+                <td
+                  :key="key"
+                  :class="cssClass"
+                  class="gobierto-table__td"
+                >
+                  <span>
+                    <a href="#">{{ item[field] }}</a>
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: empty link? what's that?

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 11]
```diff
@@ -0,0 +1,240 @@
+<template>
+  <div
+    data-testid="table-component"
+    class="gobierto-table"
+  >
+    <div class="gobierto-table__header">
+      <slot name="title" />
+      <slot
+        name="columns"
+        :toggle-visibility="toggleVisibility"
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: How this work? A slot cannot receive props, indeed, it's nonsense

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+              v-for="{key, field, type, cssClass} in filterColumns"
+            >
+              <template v-if="type === 'money'">
+                <td
+                  :key="key"
+                  :class="cssClass"
+                  class="gobierto-table__td"
+                >
+                  <span>
+                    {{ money(item[field]) }}
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: Use better a filter
```suggestion
                    {{ item[field] | money }}
```

File: app/javascript/lib/vue-components/modules/TableColumnsSelector.vue [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    },
+    showColumns: {
+      type: Array,
+      default: () => []
+    }
+  },
+  data() {
+    return {
+      showColumnsModal: false,
+      labelCustomizeColumns: I18n.t("gobierto_plans.plan_types.show.customize_columns") || ''
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/TableColumnsSelector.vue]: A component with translations binded to gobierto_plans? No way

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+          </th>
+        </template>
+      </thead>
+      <tbody>
+        <template
+          v-for="(item, index) in dataTable"
+        >
+          <router-link
+            :key="index"
+            :to="{ name: routingMember, params: {id: item[routingId] } }"
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: What if I don't want to make the cells enroutable? What if `item` has no `routingId`? This template cannot be like that

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+                  :key="key"
+                  :class="cssClass"
+                  class="gobierto-table__td"
+                >
+                  <span>
+                    <a href="#">{{ item[field] }}</a>
+                  </span>
+                </td>
+              </template>
+              <template v-else-if="type === 'truncate'">
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue:69]: What's a truncate cell type? Indeed, what's the difference with the default one? It's just wrapped with an `span`?

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+      dataTable: [],
+      containerPagination: this.paginationId
+    };
+  },
+  computed: {
+    tmpRows() {
+      return this.data || []
+    },
+    rowsSorted() {
+      const id = this.currentSortColumn.field || this.orderColumn;
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: `this.currentSortColumn` is an String by default. Then it becomes an object? What's that?

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+      const columns = this.columns.map(({ field }) => field)
+      const columnName = columns[columns.findIndex((x, index) => index === id)]
+      if (value) {
+        this.visibleColumns.push(columnName)
+      } else {
+        this.visibleColumns = this.visibleColumns.filter(item => item !== columnName)
+      }
+      this.prepareTable()
+      this.$emit('update-show-columns', this.visibleColumns)
+    },
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: This cannot be there. `toggleVisibility` is not something that Table.vue manages, but the ColumnSelector

**Review by Crashillo — CHANGES_REQUESTED** [review]
IMO, this is not a reusable component. It's just a sugar code.

**Comment by Crashillo** [comment]
Conflicts pending

**Comment by jorgeatgu** [comment]
> Conflicts pending

Ready to review

File: app/assets/stylesheets/comp-vue-table.scss [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        white-space: nowrap;
+        margin-right: .5rem;
+      }
+    }
+
+    &__icon {
+      i {
+        margin-right: .5rem;
+      }
+    }
```
Comment (Crashillo) [inline app/assets/stylesheets/comp-vue-table.scss]: Nesting

File: app/javascript/lib/vue-components/modules/TableColumnsSelector.vue (hunk 13 above)
Comment (Crashillo) [reply app/javascript/lib/vue-components/modules/TableColumnsSelector.vue]: Still pending

File: app/javascript/gobierto_visualizations/webapp/containers/subsidies/Summary.vue [hunk 19]
```diff
@@ -155,16 +155,20 @@
         {{ labelMainBeneficiaries }}
       </h3>
       <Table
-        :items="items"
-        :columns="columns"
+        :data="items"
+        :order-column="'name'"
+        :columns="grantedColumns"
+        :show-columns="showColumns"
+        :row-clickable="false"
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/webapp/containers/subsidies/Summary.vue]: Remove this prop

File: app/javascript/gobierto_visualizations/webapp/containers/contracts/ContractsIndex.vue [hunk 20]
```diff
@@ -1,13 +1,19 @@
...
-    :columns="columns"
-  />
+  <div>
+    <Table
+      :data="items"
+      :order-column="'final_amount_no_taxes'"
+      :columns="contractsColumns"
+      :show-columns="showColumns"
+      :on-row-click="goesToItem"
+      :row-clickable="true"
```
Comment (Crashillo) [inline app/javascript/gobierto_visualizations/webapp/containers/contracts/ContractsIndex.vue]: Remove this prop

File: app/javascript/lib/vue-components/modules/Table.vue [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+      default: true
+    },
+    rowClickable: {
+      type: Boolean,
+      default: true
+    },
+    onRowClick: {
+      type: Function,
+      default: x => x
+    }
```
Comment (Crashillo) [inline app/javascript/lib/vue-components/modules/Table.vue]: It's redundant declaring `onRowClick` as a function, and then, add another prop `rowClickable` to set the styles.

A row should be clickable (cursor), when `onRowClick` has any value. So, first off, remove the `rowClickable` prop. Then, set `default: null` for onRowClick. And lastly, create a computed prop who checks whether onRowClick is defined or not. If so, you may something like:

```vue
<tr
  :key="index"
  :class="{ 'is-clickable': isRowClickable }"
  class="gobierto-table__tr"
  @click="isRowClickable ? onRowClick(item) : null"
>
```
```vue
computed: {
  isRowClickable() {
    return !!this.onRowClick
  }
}
```

**Review by Crashillo — CHANGES_REQUESTED** [review]

**Review by Crashillo — APPROVED** [review]

**Comment by ferblape** [comment]
When I add multiple columns and scroll to the right, the link to configure columns moves to the left, while it should stay fixed in the right of the menu. Is it easy to fix?

Screenshot from 2021-02-05 12-38-17

**Comment by jorgeatgu** [comment]
> When I add multiple columns and scroll to the right, the link to configure columns moves to the left, while it should stay fixed in the right of the menu. Is it easy to fix?
> 
> Screenshot from 2021-02-05 12-38-17

Fixed. Now the column selector is sticky while scrolling the table.
Screen Recording 2021-02-05 at 13 43 19 2021-02-05 13_45_42

**Pull request merged**